### PR TITLE
KNCK-3255 - cloud search client in goamz

### PR DESCRIPTION
**Pull request by nalin-humin**
JIRA link: https://humindevelopment.atlassian.net/browse/KNCK-3255

Reviewers: @dosheawork, @nick-at-humin

### CODE REVIEW

File: cloudsearch/cloudsearch.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	uri    string
+}
+
+type Cloudsearch interface {
+	AddUri(string)
+	Upload([]byte) ([]byte, error)
+	Search(string) ([]byte, error)
+}
+
+func New(auth aws.Auth, region aws.Region) Cloudsearch {
```
Comment (dosheawork) [inline cloudsearch/cloudsearch.go]: Why can you construct a client without a search domain?

File: cloudsearch/cloudsearch.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+)
+
+type csClient struct {
+	auth   aws.Auth
+	region aws.Region
+	uri    string
+}
+
+type Cloudsearch interface {
+	AddUri(string)
```
Comment (dosheawork) [inline cloudsearch/cloudsearch.go]: Is the idea to share a single Cloudsearch instance between multiple search domains?

Comment (nalin-humin) [inline cloudsearch/cloudsearch.go]: yes, we may want only one client but different endpoints.

Comment (nalin-humin) [inline cloudsearch/cloudsearch.go]: as discussed, i have changed it and exposed separate client for each endpoint.

File: cloudsearch/cloudsearch.go (hunk 1 above)
Comment (nalin-humin) [inline cloudsearch/cloudsearch.go]: done

**Comment by nalin-humin** [comment]
added test client support - for testing upload/search. 2ab3b90

File: cloudsearch/cloudsearch.go [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	if hashIndex > 0 {
+		codeStr = codeStr[hashIndex+1:]
+	}
+	ddbError.Code = codeStr
+
+	return &ddbError
+}
+
+func (cs *csClient) Upload(batch_data *UploadBatchData) error {
+	return cs.save(batch_data.Data, true)
```
Comment (dosheawork) [inline cloudsearch/cloudsearch.go]: check for nil?

File: cloudsearch/cloudsearch.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	hashIndex := strings.Index(codeStr, "#")
+	if hashIndex > 0 {
+		codeStr = codeStr[hashIndex+1:]
+	}
+	ddbError.Code = codeStr
+
+	return &ddbError
+}
+
+func (cs *csClient) Upload(batch_data *UploadBatchData) error {
```
Comment (dosheawork) [inline cloudsearch/cloudsearch.go]: Why not pass the slice directly?

File: cloudsearch/cloudsearch.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+	url := fmt.Sprintf("https://%s/%s", cs.endpoint, uploadVersionedPath)
+	reader := bytes.NewReader(raw)
+
+	hreq, err := http.NewRequest("POST", url, reader)
+	if err != nil {
+		return err
+	}
+
+	_, err = cs.send(hreq, sign_require)
```
Comment (dosheawork) [inline cloudsearch/cloudsearch.go]: if there is an error do you want to log the body?

File: cloudsearch/cloudsearch_testsupport.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+	Data []DataHolder `json:"Data"`
+}
+
+type DataHolder struct {
+	Id     string            `json:"Id"`
+	Fields map[string]string `json:"Fields"`
+}
+
+const (
+	tempJsonFileName = "/var/tmp/cs_data_holder.json"
```
Comment (dosheawork) [inline cloudsearch/cloudsearch_testsupport.go]: no file required, use an in memory hash.

File: cloudsearch/cloudsearch_testsupport.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	results := make([]Result, len(cst.data_holders.Data))
+	for idx, hit := range cst.data_holders.Data {
+		results[idx] = NewSearchResult(hit.Id, hit.Fields)
+	}
+
+	return results, nil
+}
+
+func (cst *csTestClient) Delete(batch_data *DeleteBatchData) error {
+	return nil
```
Comment (dosheawork) [inline cloudsearch/cloudsearch_testsupport.go]: delete can remove from the hash.

File: cloudsearch/query_builder.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+func (q *query) Expression() string {
+	return q.expression
+}
+
+func (q *query) AddExpression(expression string) {
+	q.expression = expression
+}
+
+func (q *query) AddSize(size int64) {
```
Comment (dosheawork) [inline cloudsearch/query_builder.go]: lets add a comment on what the size is.

**Comment by dosheawork** [comment]
:+1: LGTM

**Pull request merged**